### PR TITLE
MM-1358, MM-1203 Adding preview of style guide and ESLint Javascript + JSX static analyzer

### PR DESCRIPTION
**Pull request by crspeller**
This is a preview of the upcoming style guide which will begin to be enforced after the next release. 

ESLint has been added to the build process. ESLint will be the authority on style issues. You can run it my using `make check`. However, because we currently have many errors, it would be much easier to run it on a file to file basis. The easiest way to do this is to integrate it with your editor. (see below)

## Using ESLint with your editor

1. Install eslint globally using npm: `npm i -g eslint`
2. Install the eslint react plugin: `nom i -g eslint-plugin-react`
3. Follow the instructions to integrate with your favourite editor: http://eslint.org/docs/user-guide/integrations

## Rollout plans
- Preview of style guide and ESLint. (this PR)
- Integrate feedback from style guide preview to create final version. 
- 0.6.0 is released.
- Update of dependencies and a switch from reactify to babelify.
- A few weeks to update code gradually to meet style guide. 
- One final commit to enable strict mode, enable enforcement of ESLint on Travis CI, and clean up any remaining issues.

### CODE REVIEW

File: STYLE-GUIDE.md [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+    propertyTwo="2"
+>
+  <Child />
+</Tag>
+
+// Correct
+<Tag propertyOne="1" />
+```
+
+### Nameing
```
Comment (coreyhulen) [inline STYLE-GUIDE.md]: Naming

**Comment by coreyhulen** [comment]
+1

File: STYLE-GUIDE.md [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+## Go
+
+All go code must follow the golang official [Style Guide](https://golang.org/doc/effective_go.html)
+
+In addition all code must be run though the official go formater tool [gofmt](https://golang.org/cmd/gofmt/)
+
+
+## Javascript
+
+Part of the buld process is running ESLint. ESLint is the final athority on all style issues. PRs will not be accepted unless there are no errors or warnings running ESLint. The ESLint configuration file can be found in: [web/react/.eslintrc](https://github.com/mattermost/platform/blob/master/web/react/.eslintrc.json)
```
Comment (jwilander) [inline STYLE-GUIDE.md]: athority -> authorith

File: STYLE-GUIDE.md [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+
+Instructions on how to use ESLint with your favourite editor can be found here: [http://eslint.org/docs/user-guide/integrations](http://eslint.org/docs/user-guide/integrations)
+
+The following is an abriged version of the [Airbnb Javascript Style Guide](https://github.com/airbnb/javascript/blob/master/README.md#airbnb-javascript-style-guide-), with modifications. Anything that is unclear here follow that guide. If there is a conflict, follow what is said below.
```
Comment (jwilander) [inline STYLE-GUIDE.md]: abriged -> abridged

File: STYLE-GUIDE.md [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+{
+    stuff...;
+}
+else
+{
+    stuff...;
+}
+
+// Incorrect
+if (somthing) stuff...;
```
Comment (jwilander) [inline STYLE-GUIDE.md]: The guideline says "Braces must be used on all multi-line blocks". But this isn't a multi-line block, so is it really incorrect? Personally, I like this type of one-line if for short stuff but I'm 0/5 on whether it's correct or not. If we do want it to be incorrect, then maybe update the guideline to reflect that?

File: STYLE-GUIDE.md [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+
+// Incorrect
+function wrongGetStr(stuff) {
+    return "This is the " + stuff + " string";
+}
+```
+
+## React-JSX
+
+Part of the buld process is running ESLint. ESLint is the final athority on all style issues. PRs will not be accepted unless there are no errors or warnings running ESLint. The ESLint configuration file can be found in: [web/react/.eslintrc](https://github.com/mattermost/platform/blob/master/web/react/.eslintrc.json)
```
Comment (jwilander) [inline STYLE-GUIDE.md]: buld -> build

File: STYLE-GUIDE.md [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+var x = 1;
+
+// Incorrect
+var x = 1
+```
+
+### Variables
+
+- Declarations must always use var, let or const.
+- Perfer let or const over var.
```
Comment (jwilander) [inline STYLE-GUIDE.md]: Perfer -> Prefer

Comment (jwilander) [inline STYLE-GUIDE.md]: Also do we want to enforce camel case or snake case for variable names in javascript? We're doing both all over the place. We can discuss/add this later, just wanted to bring it up

Comment (crspeller) [inline STYLE-GUIDE.md]: It is enforced by ESLint `"camelcase": [2, {"properties": "always"}]` I will add a note here about it.

File: STYLE-GUIDE.md (hunk 4 above)
Comment (crspeller) [inline STYLE-GUIDE.md]: Having braces for single line blocks is enforced by ESLint. I will add a note here clarifying.

Comment (jwilander) [inline STYLE-GUIDE.md]: Ok cool

**Comment by jwilander** [comment]
+1 other than my spelling comments

File: STYLE-GUIDE.md (hunk 6 above)
Comment (jwilander) [inline STYLE-GUIDE.md]: Cool beans

**Comment by crspeller** [comment]
Clarified based on jwilander's notes. Fixed spelling mistakes. (this is what happens when I forget to type `:set spell`)

**Comment by ralder** [comment]
:sunny: :+1:

**Pull request merged**